File: RustNet/dotnet/src/RustNet.Buses/OneWire.cs
using RustNet.Core;

namespace RustNet.Buses;

/// <summary>
/// Dallas/Maxim 1-Wire bus master. Timing-critical bit banging happens in
/// the Rust HAL; managed code works at the byte/ROM level.
/// </summary>
public static class OneWire
{
    /// <summary>Bus reset; true when at least one slave answered presence.</summary>
    [InternalCall]
    public static bool Reset(int bus) => throw new RuntimeOnlyException();

    /// <summary>Raw ROM list, 8 bytes per device little-endian; prefer <see cref="Search"/>.</summary>
    [InternalCall]
    public static byte[] SearchRaw(int bus) => throw new RuntimeOnlyException();

    /// <summary>Address one slave (MATCH ROM).</summary>
    [InternalCall]
    public static void Select(int bus, long rom) => throw new RuntimeOnlyException();

    /// <summary>Address all slaves (SKIP ROM).</summary>
    [InternalCall]
    public static void Skip(int bus) => throw new RuntimeOnlyException();

    [InternalCall]
    public static void Write(int bus, byte[] data) => throw new RuntimeOnlyException();

    [InternalCall]
    public static byte[] Read(int bus, int count) => throw new RuntimeOnlyException();

    /// <summary>Enumerate slave ROM codes on the bus.</summary>
    public static long[] Search(int bus)
    {
        byte[] raw = SearchRaw(bus);
        long[] roms = new long[raw.Length / 8];
        for (int i = 0; i < roms.Length; i++)
        {
            long rom = 0;
            for (int b = 7; b >= 0; b--)
            {
                rom = (rom << 8) | raw[i * 8 + b];
            }
            roms[i] = rom;
        }
        return roms;
    }

    public static void WriteByte(int bus, int value)
    {
        byte[] one = new byte[1];
        one[0] = (byte)value;
        Write(bus, one);
    }
}

File: RustNet/runtime/rustnet-hal/src/onewire.rs

#[allow(unused_imports)]
use alloc::{boxed::Box, string::{String, ToString}, vec::Vec};
use crate::HalResult;

/// Dallas/Maxim 1-Wire bus master.
///
/// Timing-critical bit banging lives below this trait (chip crates use a
/// hardware peripheral or cycle-counted GPIO; the simulator is
/// instantaneous), so drivers written against it are portable.
pub trait OneWireBus: Send {
    /// Bus reset. Returns true when at least one slave answered presence.
    fn reset(&mut self) -> HalResult<bool>;
    fn write_byte(&mut self, byte: u8) -> HalResult<()>;
    fn read_byte(&mut self) -> HalResult<u8>;
    fn write(&mut self, data: &[u8]) -> HalResult<()> {
        for b in data {
            self.write_byte(*b)?;
        }
        Ok(())
    }
    fn read(&mut self, buf: &mut [u8]) -> HalResult<()> {
        for b in buf.iter_mut() {
            *b = self.read_byte()?;
        }
        Ok(())
    }
    /// Enumerate all slave ROM codes on the bus (SEARCH ROM).
    fn search(&mut self) -> HalResult<Vec<u64>>;
    /// Address one slave (MATCH ROM + 8 ROM bytes).
    fn select(&mut self, rom: u64) -> HalResult<()>;
    /// Address all slaves at once (SKIP ROM).
    fn skip(&mut self) -> HalResult<()>;
}

/// Dallas CRC-8 (poly 0x31 reflected = 0x8C), used by ROM codes and
/// DS18B20 scratchpads.
pub fn crc8(data: &[u8]) -> u8 {
    let mut crc = 0u8;
    for byte in data {
        let mut b = *byte;
        for _ in 0..8 {
            let mix = (crc ^ b) & 0x01;
            crc >>= 1;
            if mix != 0 {
                crc ^= 0x8C;
            }
            b >>= 1;
        }
    }
    crc
}
